apps/wakunode2/app.nim

```diff
@@ -88,74 +85,43 @@ func node*(app: App): WakuNode =
 func version*(app: App): string =
   app.version
 
-
 ## Initialisation
 
-proc init*(T: type App, rng: ref HmacDrbgContext, conf: WakuNodeConf): T =
-  let key =
-    if conf.nodeKey.isSome():
-      conf.nodeKey.get()
-    else:
-      let keyRes = crypto.PrivateKey.random(Secp256k1, rng[])
-
-      if keyRes.isErr():
-        error "failed to generate key", error=keyRes.error
-        quit(QuitFailure)
-
-      keyRes.get()
-
-  let netConfigRes = networkConfiguration(conf, clientId)
-
-  let netConfig =
-    if netConfigRes.isErr():
-      error "failed to create internal config", error=netConfigRes.error
-      quit(QuitFailure)
-    else: netConfigRes.get()
-
-  let recordRes = enrConfiguration(conf, netConfig, key)
-
-  let record =
-    if recordRes.isErr():
-      error "failed to create record", error=recordRes.error
-      quit(QuitFailure)
-    else: recordRes.get()
-
-  if isClusterMismatched(record, conf.clusterId):
-    error "cluster id mismatch configured shards"
-    quit(QuitFailure)
-
-  App(
-    version: git_version,
-    conf: conf,
-    netConf: netConfig,
-    rng: rng,
-    key: key,
-    record: record,
-    node: nil
-  )
-
-proc setupPeerPersistence*(app: var App): AppResult[void] =
-  if not app.conf.peerPersistence:
-    return ok()
-
-  let peerStoreRes = setupPeerStorage()
-  if peerStoreRes.isErr():
-    return err("failed to setup peer store" & peerStoreRes.error)
-
-  app.peerStore = peerStoreRes.get()
-
-  ok()
-
-proc setupDyamicBootstrapNodes*(app: var App): AppResult[void] =
-  let dynamicBootstrapNodesRes = retrieveDynamicBootstrapNodes(app.conf.dnsDiscovery,
-                                                               app.conf.dnsDiscoveryUrl,
-                                                               app.conf.dnsDiscoveryNameServers)
-  if dynamicBootstrapNodesRes.isOk():
-    app.dynamicBootstrapNodes = dynamicBootstrapNodesRes.get()
-  else:
-    warn "2/7 Retrieving dynamic bootstrap nodes failed. Continuing without dynamic bootstrap nodes.", error=dynamicBootstrapNodesRes.error
-
-  ok()
+proc init*(T: type App, conf: WakuNodeConf): Result[App, string] =
+
+  var confCopy = conf
+  let rng = crypto.newRng()
+
+  if not confCopy.nodekey.isSome():
+    let keyRes = crypto.PrivateKey.random(Secp256k1, rng[])
+    if keyRes.isErr():  
+      error "Failed to generate key", error = $keyRes.error
+      return err("Failed to generate key: " & $keyRes.error)
+    confCopy.nodekey = some(keyRes.get())
+
+  debug "Retrieve dynamic bootstrap nodes"
+  let dynamicBootstrapNodesRes = retrieveDynamicBootstrapNodes(confCopy.dnsDiscovery,
+                                                              confCopy.dnsDiscoveryUrl,
+                                                              confCopy.dnsDiscoveryNameServers)
+  if dynamicBootstrapNodesRes.isErr():
```

Comment:
same as above

Reply:
If I remember correctly I initially tried `valueOr()` but for some reason had compilation errors related to logging that I didn't have anywhere else - so just used the other approach

I'll try again to double check :)

Reply:
Yap, it was that.

For example if I change it to

```
let dynamicBootstrapNodes = retrieveDynamicBootstrapNodes(confCopy.dnsDiscovery,
                                                              confCopy.dnsDiscoveryUrl,
                                                              confCopy.dnsDiscoveryNameServers).valueOr:
    error "Retrieving dynamic bootstrap nodes failed", error = $error
    return err("Retrieving dynamic bootstrap nodes failed: " & $error)
```
    
It has issues printing the variable `error` because it thinks it's of type `None`
    
Not sure why, as the return value of `retrieveDynamicBootstrapNodes()` is `Result[seq[RemotePeerInfo], string]`

I think that after some attempts trying to figure it out I just wrote it the other way and moved forward.

Any ideas?

Reply:
Yeah I too noticed some edge cases (I hate those). If only the compiler was smarter... *cough itwouldworkinrust cough* oh well :shrug:

Reply:
loool (but true) 😅 😵